### ISSUE TITLE
Implementation of p-dimension learning rate

### ISSUE THREAD

**Issue by k-ye** [airoldilab/sgd#12]
Hi everyone~

I've implemented **p-dimension** learning rate just now, with the source code in `learnrate` branch. Currently, it supports the user to either pass in a string ('uni-dim' or 'px-dim') or an interger({1, 2}) to specify what kind of learning rate they prefer to use in our main interface.

Several issues I've encountered for now:
- @lantian2012  and I had a really long discussion about what we should pass in to the learning rate function to make it generic enough. (Before we are only passing in _t_, which represents the current number of iteration. Soon we discovered that this was not enough for _p-dimension_ case.) Right now what we passed in was the t-th datapoint, the previous _theta_ and _t_ itself.
- This p-dimension learning rate did not perform that well as the previous one did. I still used the system generated poisson dataset to run the test. The true theta should be around (0.7, 1.4), but what we got when using p-dimension learning rate and 'asgd' method was (0.52, 1.29), and (0.16, 1.45) with 'implicit'. Here's the code of the learning rate:

```
static mat learning_rate(const mat& theta_old, const Imp_DataPoint& data_pt, 
                          unsigned t, unsigned p,
                          score_func_type score_func) {
    mat Idiag(p, p, fill::eye);
    mat Gi = score_func(theta_old, data_pt);
    Idiag = Idiag + diagmat(Gi * Gi.t());

    for (unsigned i = 0; i < p; ++i) {
      if (abs(Idiag.at(i, i)) > 1e-8) {
        Idiag.at(i, i) = 1. / Idiag.at(i, i);
      }
    }
    return Idiag;
  }
```

Get any idea where might went wrong...? (One thing that confused me was: is this `Idiag` matrix initialized just once at the very beginning or in each iteration step?)
- In #8 where @ptoulis mentioned **Auto-full** method, could you please offer some more information on what that is? Thanks :)
